pkg/manifest/locator.go

```diff
@@ -0,0 +1,190 @@
+	progression := parseOptFloat64(rawJson["progression"])
+	if progression >= 0.0 && progression <= 1.0 {
+		locations.Progression = progression
+	}
+
+	// Position
+	position := parseOptFloat64(rawJson["position"])
+	if position > 0 {
+		locations.Position = float64ToUint(position)
+	}
+
+	// TotalProgression
+	totalProgression := parseOptFloat64(rawJson["totalProgression"])
+	if totalProgression >= 0.0 && totalProgression <= 1.0 {
+		locations.TotalProgression = totalProgression
+	}
+
+	// Delete above vals so that we can put everything else in OtherLocations
+	for _, v := range []string{
+		"fragments", "fragment", "progression", "position", "totalProgression",
+	} {
+		delete(rawJson, v)
+	}
+
+	// Now all we have left is everything else!
+	if len(rawJson) > 0 {
+		locations.OtherLocations = rawJson
+	}
+
+	return locations, nil
+}
+
+func (l *Locations) UnmarshalJSON(b []byte) error {
+	var object map[string]interface{}
+	err := json.Unmarshal(b, &object)
+	if err != nil {
+		return err
+	}
+	fl, err := LocationsFromJSON(object)
+	if err != nil {
+		return err
+	}
+	*l = *fl
+	return nil
+}
+
+func (l Locations) MarshalJSON() ([]byte, error) {
+	j := l.OtherLocations
+	if j == nil {
+		j = make(map[string]interface{})
+	}
+
+	if len(l.Fragments) > 0 {
+		j["fragments"] = l.Fragments
+	}
+	if l.Progression != 0.0 {
+		j["progression"] = l.Progression
+	}
+	if l.Position > 0 {
+		j["position"] = l.Position
+	}
+	if l.TotalProgression > 0 {
+		j["totalProgression"] = l.TotalProgression
+	}
+
+	return json.Marshal(j)
+}
+
+// Textual context of the locator.
+// A Locator Text Object contains multiple text fragments, useful to give a context to the [Locator] or for highlights.
+// https://github.com/readium/architecture/tree/master/models/locators#the-text-object
+type Text struct {
+	Before    string `json:"before,omitempty"`    // The text before the locator.
+	Highlight string `json:"highlight,omitempty"` // The text at the locator.
+	After     string `json:"after,omitempty"`     // The text after the locator.
+}
+
+func TextFromJSON(rawJson map[string]interface{}) *Text {
+	if rawJson == nil {
+		return nil
+	}
+
+	return &Text{
+		Before:    parseOptString(rawJson["before"]),
+		Highlight: parseOptString(rawJson["highlight"]),
+		After:     parseOptString(rawJson["after"]),
+	}
+}
+
+/**
+ * Provides a precise location in a publication in a format that can be stored and shared.
+ *
+ * There are many different use cases for locators:
+ *  - getting back to the last position in a publication
+ *  - bookmarks
+ *  - highlights & annotations
+ *  - search results
+ *  - human-readable (and shareable) reference in a publication
+ *
+ * https://github.com/readium/architecture/tree/master/locators
+ */
+type Locator struct {
+	Href      string     `json:"href"`
+	Type      string     `json:"type"`
+	Title     string     `json:"title,omitempty"`
+	Locations *Locations `json:"locations,omitempty"`
+	Text      *Text      `json:"text,omitempty"`
```

Comment:
```suggestion
	Locations Locations `json:"locations,omitempty"`
	Text      Text      `json:"text,omitempty"`
```

Reply:
Could you explain why you are suggesting this? According to https://github.com/readium/architecture/blob/master/models/locators/README.md#the-locator-object the locations and text properties are optional, and therefore the nicest JSON you can get would omit these properties if they have no value. If they aren't pointers, but direct values, then they will be the struct with the fields initialized to their default values. So, for example, a position object would go from having no `text` property, to having `text: {}`

Reply:
The rational was:

* Safer as the locations and text can't be mutated from outside of the current scope on a `Locator` instance you hold.
* Easier to use when manipulating a `Locator`, because you don't need to check whether `Locations` or `Text` are `null` when accessing their properties. Instead you can do directly `locator.Locations.Progression`, or directly set `locator.Text.Before = "before"`. This will be particularly relevant when implementing APIs such as the search service.

Having empty JSON objects would be alright I guess. I'd prefer if we keep the models safe and not impacted by [the shortcoming of Go's serialization library](https://go-review.googlesource.com/c/go/+/13914/).